### PR TITLE
ensure user details are always in the session after entering password

### PR DESCRIPTION
**Pull request by leohemsted**
We signal that we're mid-way through the sign-in flow by adding a `user_details` dict to the session.

previously, we'd only put a user's details in the session in `User.sign_in`, just before sending any 2fa prompt and redirecting to the two factor pages.

However, we found a bug where a user with no session (eg, using a fresh browser) tried to log in, but they had never clicked the link to validate their email address when registering. Their user's state was still in "pending", so we redirected to `main.resend_email_verification` as intended - however, they didn't have anything in the session and the resend page expected to get the email address to resend to out of that.

To be safe, as soon as we've confirmed the user has entered their password correctly, lets save the session data at that point. That way any redirects will be fine.

### CODE REVIEW

**Review by benthorner — APPROVED** [review]
Nice catch, and very fiddly code to change 🎖️.

It took me a while to work out what the sequence was to get to `resend_email_verification`:

- A pending user hits the `/register` route, which calls `_do_registration`.
- `_do_registration` populates `user_details` in the session, then calls `send_verify_email`.
- A user normally clicks on the link in that email, so they hit `/verify-email/<token>`.
- Our user doesn't. When they go to Notify in a separate browser, they hit `/sign-in` instead.
- `/sign-in` tries to redirect them to `resend_email_verification`, which has `@redirect_to_sign_in`.
- So the user is caught in a loop.

File: app/main/views/sign_in.py [hunk 1]
```diff
@@ -33,24 +33,31 @@ def sign_in():
...
-            if user.sms_auth:
-                return redirect(url_for('.two_factor_sms', next=redirect_url))
-            if user.email_auth:
-                return redirect(url_for('.two_factor_email_sent', next=redirect_url))
-            if user.webauthn_auth:
-                return redirect(url_for('.two_factor_webauthn', next=redirect_url))
+            if user.state == 'pending':
+                return redirect(url_for('main.resend_email_verification', next=redirect_url))
+
+            if user.is_active:
```
[the remark this answers is not in the document]
Comment (leohemsted) [reply app/main/views/sign_in.py:43]: hmmm, this is a great point - i'm not sure this is a typical expected flow.

in `accept_invite`, we set `invited_user_id` in accept_invite (provided invite is valid and still available). We generally keep this in session until the user gets to (any) service dashboard 

If you're already signed in, when you click the invite link we handle clashes with email addresses there.

If you're not signed in, you get redirect to `register_from_invite`. This won't take you to sign_in though, it'll take you to `verify` (which is a separate but similar page related to the register flow instead). I tested this locally.
```127.0.0.1 - - [30/Jun/2021 17:49:25] "GET /invitation/ImNhZDBiNjg2LTcwNWUtNDFlMS04NGI1LTAwOWU2ZWRjZTRkYiI.YNw1Gg.nkTlsMuOtZEYRLpWMsY6Wtrnst8 HTTP/1.1" 302 -
127.0.0.1 - - [30/Jun/2021 17:49:25] "GET /register-from-invite HTTP/1.1" 200 -
127.0.0.1 - - [30/Jun/2021 17:49:35] "POST /register-from-invite HTTP/1.1" 302 -
127.0.0.1 - - [30/Jun/2021 17:49:35] "GET /verify HTTP/1.1" 200 -
127.0.0.1 - - [30/Jun/2021 17:49:44] "POST /verify HTTP/1.1" 302 -
127.0.0.1 - - [30/Jun/2021 17:49:44] "GET /services/d0829bc4-801d-4d54-ab38-baadf61bfe0b HTTP/1.1" 200 -```

There may be weird edge cases if you manually go to the sign_in url after entering your password but before verifying sms code but i really cant figure any way you'd get there through normal interactions.

[the remark this answers is not in the document]
Comment (benthorner) [reply app/main/views/sign_in.py:43]: Do you think we can remove the conditional then?

I tried without and all the tests pass, so it seems to be more trouble (i.e. confusing) than it's worth.

[the remark this answers is not in the document]
Comment (leohemsted) [reply app/main/views/sign_in.py:43]: okay i just figured out how you get this.

Someone sends an invite to user A, who already exists in notify. 

When a signed out browser opens that link, it goes to the sign in page. this flow is documented (somewhat poorly) by the test  `test_accept_invite.py::test_existing_signed_out_user_accept_invite_redirects_to_sign_in`

Then, if you enter user B's sign in details. it compares your email to the invited user's email, and if not, complains.

I think this is a real situation that could conceivably happen, especially if someone moves depts, is invited to their new notify team, but then they muscle memory type in their old email address.

-----------------------

In conclusion:

the user.active check should stay, and it's fine for it to happen earlier. If the user is inactive then i don't really mind what happens as long as we don't progress the sign-in flow.

the email invited_user check needs to stay.

i've added a negative test for the invited user (make sure it accepts the invite if you sign in with a case insensitive same email) but not a positive test (make sure it rejects if you sign in as someone completely different). i've added that test now.

[the remark this answers is not in the document]
Comment (benthorner) [reply app/main/views/sign_in.py:43]: Thanks, that makes sense when I re-read what you said before:

> If you're already signed in ...
> 
> If you're not signed in, you get redirect to register ...

☝️ there's a logical gap for "not signed in, but already registered", which is what you've covered here for the invite code.

Going back to the `is_active` check, I found I can remove it and all the tests still pass. We do have [a test for that scenario](https://github.com/alphagov/notifications-admin/blob/256c840b46d4095edbd0af105e5c1a340242575d/tests/app/main/views/test_sign_in.py#L191), but it doesn't check what's _not_ done in this block. We're generally covered by Flask-Login `is_active`, but I thought it was worth mentioning in case you want to extend that test.

**Review by benthorner — APPROVED** [review]
Nice detective work 🕵️.
